fea-rs/src/compile/compiler.rs

```diff
@@ -0,0 +1,157 @@
+//! The main public API for compilation
+
+use std::{
+    ffi::OsString,
+    path::{Path, PathBuf},
+};
+
+use crate::{
+    parse::{FileSystemResolver, SourceResolver},
+    Diagnostic, GlyphMap, ParseTree,
+};
+
+use super::{
+    error::{CompilerError, DiagnosticSet},
+    Compilation, Opts,
+};
+
+/// A builder-style entry point for the compiler.
+///
+/// This is intended as the principal public API for this crate.
+///
+/// ```no_run
+/// # use fea_rs::Compiler;
+/// # fn make_glyph_map() -> fea_rs::GlyphMap { todo!() }
+/// let glyph_map = make_glyph_map();
+/// let my_font_bytes = Compiler::new("path/to/features.fea", &glyph_map)
+///     .verbose(true)
+///     .compile_binary().unwrap();
+/// ```
+#[derive(Debug)]
+pub struct Compiler<'a> {
+    root_path: OsString,
+    project_root: Option<PathBuf>,
+    glyph_map: &'a GlyphMap,
+    verbose: bool,
+    opts: Opts,
+    resolver: Option<Box<dyn SourceResolver>>,
+}
+
+impl<'a> Compiler<'a> {
+    /// Configure a new compilation run with a root source and a glyph map.
+    ///
+    /// In the general case, `root_path` will be a path to a feature file on disk;
+    /// however you may compile from memory by passing a custom [`SourceResolver`]
+    /// to the [`with_resolver`] method, in which case `root_path` can be any
+    /// identifier that your resolver will resolve.
+    ///
+    /// [`with_resolver`]: Self::with_resolver
+    pub fn new(root_path: impl Into<OsString>, glyph_map: &'a GlyphMap) -> Self {
+        Compiler {
+            root_path: root_path.into(),
+            glyph_map,
+            opts: Default::default(),
+            verbose: false,
+            resolver: Default::default(),
+            project_root: Default::default(),
+        }
+    }
+
+    /// Provide a custom `SourceResolver`, for mapping paths to their contents.
+    pub fn with_resolver(mut self, resolver: impl SourceResolver + 'static) -> Self {
```

Comment:
Probably user error but trying to actually populate resolver from a closure caused me fits.

Reply:
Also it's hard to fulfil `fn get_contents(&self, path: &OsStr) -> Result<String, SourceLoadError>` when there is no apparent way to construct a SourceLoadError; private fields and pub(crate) new.

Reply:
Also ... doesn't Result<String, ...> mean I have to clone the fea content? Perhaps this is a spot for the ever popular Rc?

Reply:
yes, fair, it is stored as `Rc<str>` anyway, so might as well expose that here.